File: hostd/src/api/routes/network.rs
//! Network config routes: read the VM's live [`NetworkConfig`] (which the
//! `/ports` endpoints mutate). There is intentionally no update route:
//! `exposed_ports` is managed via `/vms/{id}/ports`, and the remaining
//! fields (`allow_internet`, `egress`, `public_access`) are not enforced
//! anywhere yet, so accepting writes would pretend otherwise.

use axum::{Json, Router, extract::Path, extract::State, routing::get};

use crate::{
    api::{error::ApiResult, server::AppState},
    error::Error,
    net::NetworkConfig,
    vmm::vm::VmId,
};

/// Network routes, nested under `/vms/{id}/network`.
pub(crate) fn routes() -> Router<AppState> {
    Router::new().route("/", get(get_network))
}

/// Get the network config of a VM.
async fn get_network(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> ApiResult<Json<NetworkConfig>> {
    let instance_ref = state
        .vmm
        .get_vm(&VmId(id.clone()))
        .await?
        .ok_or(Error::VmNotFound(id))?;
    let config = instance_ref.lock()?.vm_config.network_config.clone();
    Ok(Json(config))
}

File: clients/node/src/network.ts
import { HttpClient } from './http.js';
import type { NetworkConfig } from './types.js';

/**
 * Read-only `/api/vms/{id}/network` resource: the VM's live `NetworkConfig`.
 * The `exposed_ports` list is managed via the `/ports` endpoints; the
 * remaining fields are not writable through the API.
 */
export class NetworkApi {
  /** @internal */
  constructor(
    private readonly http: HttpClient,
    private readonly vmId: string,
  ) {}

  /** Fetch the VM's current network config. */
  async get(): Promise<NetworkConfig> {
    return this.http.request<NetworkConfig>({
      method: 'GET',
      path: `/api/vms/${this.vmId}/network`,
    });
  }
}
